common/src/backoff.rs

```diff
@@ -11,12 +11,69 @@
 //! - An inaccessible network
 //! - An overloaded server
 
+use std::future::Future;
 use std::time::Duration;
+use std::time::Instant;
 
 pub use ::backoff::future::{retry, retry_notify};
 pub use ::backoff::Error as BackoffError;
 pub use ::backoff::{backoff::Backoff, ExponentialBackoff, Notify};
 
+pub trait NotifyTracking<E> {
+    fn notify_ext(&mut self, err: E, count: usize, duration: Duration);
+}
+
+impl<E, F> NotifyTracking<E> for F
+where
+    F: FnMut(E, usize, Duration),
+{
+    fn notify_ext(&mut self, err: E, count: usize, duration: Duration) {
+        self(err, count, duration)
+    }
+}
+
+/// A helper function which modifies what information is tracked within the
+/// callback of the notify function.
+///
+/// The default "Notify" function returns an error and the duration since the
+/// *last* call of notify, but often the decision about "where should an error
+/// be logged" depends on the *total* duration of time since we started
+/// retrying.
+///
+/// By returning:
+/// - A count of total calls
+/// - The duration since we started retrying
+///
+/// The caller can more easily log:
+/// - (info) Something non-erroneous on the first failure (indicating, to a log, that
+/// retry is occuring).
+/// - (warning) Something more concerned after a total time threshold has passed.
+///
+/// Identical to [::backoff::future::retry_notify], but with invokes
+/// the [NotifyTracking::notify_ext] method instead of
+/// [::backoff::Notify::notify].
+pub async fn retry_notify_ext<I, E, Fn, Fut, B, N>(
+    backoff: B,
+    operation: Fn,
+    mut notify: N,
+) -> Result<I, E>
+where
+    B: Backoff,
+    Fn: FnMut() -> Fut,
+    Fut: Future<Output = Result<I, BackoffError<E>>>,
+    N: NotifyTracking<E>,
```

Comment:
Out of curiosity more than a request for a change: Why do we have a special `NotifyTracking` trait (implemented for all `FnMut` of a certain shape) instead of taking an `FnMut` directly?

Reply:
The backoff crate defines a `Notify` trait that's part of the interface for retry_notify: https://docs.rs/backoff/0.4.0/backoff/trait.Notify.html

This trait is also implemented generically for all `FnMut` that take "error, duration since last call" as an argument.

With `NotifyTracking`, I was trying to create something similar, just with different parameters ("error, call count, total duration").

I probably *could* skip the trait and just take a `FnMut`? But, as a nitpick:
- It's technically more restrictive than the generic type here
- It's divergent from what the `backoff` crate is normally doing, which IMO makes it slightly less clear.

Reply:
I also found this a little confusing.  What about adding some of what you wrote there in a comment on the new trait?

Reply:
After thinking about this more, I think the `Notify` trait in backoff is a poor style choice, and this function would be clearer if we did not emulate it.

Just looking at this function signature, neither (a) the fact that I can pass a closure directly nor (b) the args / return type of that closure are visible: I have to go look at `NotifyTracking` and then see the `impl<F: FnMut...>` defined for it. On the other hand, if we took an `FnMut` directly, both of those become obvious, and if I had a value / concrete type that I wanted to pass here, it's trivial to pass a 1-line closure that calls into the relevant handling method.

Reply:
Fair enough, I hear the confusion, I'll switch it back. I guess the goal of aligning with the `backoff` implementation is not as important as clarity within our own codebase, and I buy that.

Updated in https://github.com/oxidecomputer/omicron/pull/2899/commits/cc8f37c7ba4ca91434dc1fb5098dec67bdd6da42 to use `FnMut` directly, plus impl traits to make it even more clear.